### PR TITLE
Bonus points

### PR DESCRIPTION
**Pull request by Larspolo**
## Description
Added the possibility for graders to add a bonus grade to a journal

# Fixes
- #294 

#### Teacher
image

#### Student
image

### CODE REVIEW

**Comment by engelhamer** [comment]
I would suggest adding the bonus part tot the student card (so above the point count and under the name) rather than its current location.

**Comment by engelhamer** [comment]
Also a question mark with tooltip would be helpful. It might not immediately become clear what the function of these points is.

**Comment by mjvkeulen** [comment]
Currently, these points magically appear to a student. I would like to be able to add a bonus node to the timeline as an educator, consisting of a rich-text description and bonus points.

We also require a mass import function, consisting of:
- username/email
- bonus points
- date
- description

**Comment by engelhamer** [comment]
Personally I am not a fan of a very extensive bonus system: a simple field out here is fine, in addition to a bulk import feature. I feel like more complicated stuff is not worthwile, since most 'bonus' points will simply be added to a particular entry anyways.

We can still add a node to a student's journal if they were granted bonus points (so a single one, all the way at the start) or include that in the start node.

File: src/vue/src/components/assets/Tooltip.vue [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<template>
```
Comment (engelhamer) [inline src/vue/src/components/assets/Tooltip.vue]: I should've done this right away :upside_down_face:

File: src/django/VLE/views/assignment.py [hunk 2]
```diff
@@ -273,6 +274,44 @@ def upcoming(self, request):
...
+                bonus = float(bonus)
+                user = User.objects.get(username=str(username))
+                journal = Journal.objects.get(assignment=assignment, user=user)
+                journal.bonus_points = bonus
+                journal.save()
+            except ValueError:
+                return response.bad_request(
+                    'The bonus file is not correctly formatted. Please use "[username],[bonuspoints]"')
+            except User.DoesNotExist:
+                return response.bad_request(
```
Comment (engelhamer) [inline src/django/VLE/views/assignment.py]: I think we should go on with parsing the file on a not found user, and simply print a list of wrong lines afterwards. Currently all points up until a wrong name are saved, but the rest is not (which seems unlogical).

Comment (mjvkeulen) [reply src/django/VLE/views/assignment.py]: I believe nothing should be saved unless the format is correct for all entries, how the error handling should take place is debatable. I'll push a suggestion, see what you guys think.

Comment (mjvkeulen) [reply src/django/VLE/views/assignment.py]: The toast will ignore string formatting, so an error output file is required. Demanding more custom upload field logic.

File: src/vue/src/views/Assignment.vue [hunk 3]
```diff
@@ -32,11 +32,20 @@
             </div>
             <b-button
                 v-if="$hasPermission('can_publish_grades') && assignmentJournals.length > 0"
-                class="add-button full-width"
+                class="add-button full-width multi-form"
                 @click="publishGradesAssignment">
                 <icon name="upload"/>
                 Publish all Grades for this Assignment
             </b-button>
+            <file-upload-input
```
Comment (mjvkeulen) [inline src/vue/src/views/Assignment.vue]: Currently a page refresh is required after both an incorrect upload (in order to clear the file upload button) or after a successful upload in order to see the bonus points applied.

**Comment by mjvkeulen** [comment]
So the toast ignores newlines and is shown too briefly. Ideally we want a file with error output. This requires some tweaks to the file upload component. We are really increasing the scope haha

**Comment by engelhamer** [comment]
A modal that shows a list (returned from the API call) would do right. I think a file would be too much of a hassle.

**Comment by engelhamer** [comment]
image

Added a modal with some more explanation. Especially the fact that points are overwritten is important to note.

**Comment by engelhamer** [comment]
Currently the server crashes if a user uploads something that is not a csv (e.g. an image). This should be fixed.

**Comment by engelhamer** [comment]
Errors are now displayed as follows.

out-11

File: src/vue/src/components/assets/file_handling/FileUploadInput.vue [hunk 4]
```diff
@@ -25,19 +25,22 @@ export default {
...
+            default: 'No file chosen'
         },
         contentID: {
             default: null
         }
     },
     data () {
         return {
-            placeholderText: 'Select a file.',
+            placeholderText: 'No file chosen',
```
Comment (Larspolo) [inline src/vue/src/components/assets/file_handling/FileUploadInput.vue]: I preferred "select a file." Why this change?

Comment (engelhamer) [reply src/vue/src/components/assets/file_handling/FileUploadInput.vue]: Because it normally displays the file name in this location once chosen. This seemed more intuitive to me.
